server/repository/src/migrations/mod.rs

```diff
@@ -0,0 +1,156 @@
+        return Err(MigrationError::DatabaseVersionAboveAppVersion(
+            database_version,
+            to_version,
+        ));
+    }
+
+    if database_version.is_pre_release() {
+        return Err(MigrationError::DatabaseVersionIsPreRelease(
+            database_version,
+        ));
+    }
+
+    for migration in migrations {
+        let migration_version = migration.version();
+
+        if migration_version > to_version {
+            // During test, when we specify to_version manually, we want migrations
+            // to stop at that version, even if there are migrations after specified to_version
+            if cfg!(test) {
+                break;
+            }
+
+            return Err(MigrationError::MigrationAboveAppVersion(
+                migration_version,
+                to_version,
+            ));
+        }
+
+        let database_version = get_database_version(connection);
+
+        // TODO transaction ?
+
+        if migration_version > database_version {
+            log::info!("Running database migration {}", migration_version);
+            migration
+                .migrate(connection)
+                .map_err(|source| MigrationError::MigrationError {
+                    source,
+                    version: migration_version.clone(),
+                })?;
+            set_database_version(connection, &migration_version)?;
+        }
+    }
+
+    set_database_version(connection, &to_version)?;
+    Ok(to_version)
+}
+
+fn get_database_version(connection: &StorageConnection) -> Version {
+    match KeyValueStoreRepository::new(connection).get_string(KeyValueType::DatabaseVersion) {
+        Ok(Some(version_str)) => Version::from_str(&version_str),
+        // Rust migrations start at "1.0.3"
+        _ => Version::from_str("1.0.3"),
+    }
+}
+
+fn set_database_version(
+    connection: &StorageConnection,
+    new_version: &Version,
+) -> Result<(), RepositoryError> {
+    KeyValueStoreRepository::new(connection)
+        .set_string(KeyValueType::DatabaseVersion, Some(new_version.to_string()))
+}
+
+#[derive(Error, Debug)]
+#[error("Sql error {0}")]
+pub(crate) struct SqlError(String, #[source] RepositoryError);
+
+/// Will try and execute diesel query return SQL error which contains debug version of SQL statements
+pub(crate) fn execute_sql_with_error<'a, Q>(
+    connection: &StorageConnection,
+    query: Q,
+) -> Result<usize, SqlError>
+where
+    Q: methods::ExecuteDsl<DBConnection>,
+    Q: QueryFragment<DBType>,
+{
+    let debug_query = diesel::debug_query::<DBType, _>(&query).to_string();
+    Q::execute(query, &connection.connection).map_err(|source| SqlError(debug_query, source.into()))
+}
+
+/// Macro will create and run SQL query, it's a less verbose way of running SQL in migrations
+macro_rules! sql {
+    ($connection:expr, $($arg:tt)*) => {{
+        let query = diesel::sql_query(&format!($($arg)*));
+        crate::migrations::execute_sql_with_error($connection, query)
```

Comment:
Is the macro needed here? What does the * do here? is it for multiple arguments?

Reply:
> What does the * do here

I've added a comment 2bf1f4e4edf3c54b682e170ec63561c47dc39414, I took it from format! macro. 

> Is the macro needed here

I wanted a single method to do the query, since it will be used very frequently, and the reason for macro was the use of `format!`